### PR TITLE
[UAVCANv1] Collection of fixes

### PR DESCRIPTION
**Pull request by JacobCrabill**
Collection of fixes for the UAVCANv1 module:
- Remove timed-out CAN frames from canard's Tx queue
- Add support to SubscriptionManager for dynamic port ID updates (instantiate and/or update the port IDs of all DynamicSubscriber objects)
- Fix the port ID update for Publishers

What has been tested (Using a Pixhawk 4):
- Updating the PX4 `UCAN1_*` parameters for all pubs/subs at run-time, with expected results
- Use `yakut monitor` to verify the expected port ID's being published

What has not yet been tested (and is being worked on separately):
- Verify that high-rate messages can be sent (O(100Hz) and larger)

### CODE REVIEW

**Comment by PetervdPerk-NXP** [comment]
LGTM, does it work on the Pixhawk 4

**Comment by JacobCrabill** [comment]
It's not working... maybe I'll commit my debug printouts.  It seems to be constantly timing out at the call to `poll()` within `CanardNuttXCDev::transmit()`

File: src/drivers/uavcan_v1/SubscriptionManager.hpp [hunk 1]
```diff
@@ -60,6 +60,7 @@
 typedef struct {
 	const char *px4_name;
 	UavcanDynamicPortSubscriber *(*create_sub)(CanardInstance &ins, UavcanParamManager &pmgr) {};
+	UavcanDynamicPortSubscriber *instance {nullptr};
```
Comment (JacobCrabill) [inline src/drivers/uavcan_v1/SubscriptionManager.hpp:63]: @PetervdPerk-NXP I'm open to other suggestions on how to handle this, but storing a pointer to the Subscriber instance was my fix for determining whether any given Subscriber was instantiated or not in future calls to `updateDynamicSubscriptions()`.

Comment (PetervdPerk-NXP) [reply src/drivers/uavcan_v1/SubscriptionManager.hpp:63]: Hm I get it why that would be faster, but I made `_uavcan_subs const` to save RAM by storing it into flash.

Another way is to iterate over the _dynsubscribers when a parameter gets updated to figure out if the subscriber already has been created, this does take more CPU cycles but saves on RAM usage, Anyway I would argue that parameter updates shouldn't occur that often e.g. only during preflight.

File: src/drivers/uavcan_v1/ParamManager.hpp [hunk 2]
```diff
@@ -47,23 +47,31 @@
...
 		if (out_val >= 0 && out_val <= CANARD_PORT_ID_MAX) {
 			out.natural16.value.elements[0] = (uint16_t)out_val;
-			out.natural16.value.count = 1;
-			uavcan_register_Value_1_0_select_natural16_(&out);
-			return true;
 
 		} else {
-			return false;
+			// "Invalid" value -- set to "UNSET"
+			out.natural16.value.elements[0] = CANARD_PORT_ID_UNSET;
```
Comment (PetervdPerk-NXP) [inline src/drivers/uavcan_v1/ParamManager.hpp:69]: UAVCAN Value1.0 has a build-in way to express it's undefined see 
https://github.com/UAVCAN/public_regulated_data_types/blob/master/uavcan/register/Value.1.0.uavcan#L6

Calling `uavcan_register_Value_1_0_select_empty_(&out);` would do the trick

File: src/drivers/uavcan_v1/Uavcan.hpp [hunk 3]
```diff
@@ -113,7 +113,7 @@ class UavcanNode : public ModuleParams, public px4::ScheduledWorkItem
 	* Current size of 8192 bytes is arbitrary, should be optimized further
 	* when more nodes and messages are on the CAN bus
 	*/
-	static constexpr unsigned HeapSize = 8192;
+	static constexpr unsigned HeapSize = 2 * 8192;
```
Comment (PetervdPerk-NXP) [inline src/drivers/uavcan_v1/Uavcan.hpp:116]: This should be configurable, either board-wise see #17100, or using a paramter. Our FMUK66 is on low on free RAM.

File: src/drivers/uavcan_v1/SubscriptionManager.hpp (hunk 1 above)
Comment (JacobCrabill) [reply src/drivers/uavcan_v1/SubscriptionManager.hpp:63]: `const` structs like this get stored in flash vs. RAM?  I guess that makes sense; didn't even consider that.  Learn something every day!

I'll try to make that change later today.

File: src/drivers/uavcan_v1/Uavcan.hpp (hunk 3 above)
Comment (JacobCrabill) [reply src/drivers/uavcan_v1/Uavcan.hpp:116]: Oh right, I didn't mean to commit this; I was playing around with code and at one point noticed I had run out of RAM so I increased it.  I'll revert this.

File: src/drivers/uavcan_v1/SubscriptionManager.hpp (hunk 1 above)
Comment (JacobCrabill) [reply src/drivers/uavcan_v1/SubscriptionManager.hpp]: Changes made

File: src/drivers/uavcan_v1/ParamManager.hpp (hunk 2 above)
Comment (JacobCrabill) [reply src/drivers/uavcan_v1/ParamManager.hpp:69]: So, are you suggesting replacing all `x != CANARD_PORT_ID_UNSET` checks with `uavcan_register_Value_1_0_is_empty_(x)`?

File: src/drivers/uavcan_v1/Uavcan.hpp (hunk 3 above)
Comment (JacobCrabill) [reply src/drivers/uavcan_v1/Uavcan.hpp]: Fixed

**Review by dagar — DISMISSED** [review]

**Comment by dagar** [comment]
There's a small merge conflict now from https://github.com/PX4/PX4-Autopilot/pull/17793.

**Comment by JacobCrabill** [comment]
Merge resolved.

One last little change I made - I kept getting confused about the `_PUB` vs. `_PID` param suffix, so I changed `_PID` to `_SUB` to make sense in relation to `_PUB`.  Can change it back if desired.

**Comment by JacobCrabill** [comment]
> What has not yet been tested (and is being worked on separately):
> 
> * Verify that high-rate messages can be sent (O(100Hz) and larger)

Note: the change to `CONFIG_CAN_FIFOSIZE=32` partially resolves the issues around transmitting / receiving high-rate, large messages using the NuttX CDev CAN driver; I'll have a follow-up PR that fully fixes it.

**Review by dagar — APPROVED** [review]

**Pull request merged**